torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1117,6 +1119,207 @@ def arg_extractor(combine_fn, input, dim):
+        ):
+            inp_meta = inp_proxy.node.meta["example_value"]
+            ini_meta = ini_proxy.node.meta["example_value"]
+            carry_meta = carry.node.meta["example_value"]
+            out_meta = out.node.meta["example_value"]
+            if (
+                carry_meta.device != inp_meta.device
+                or out_meta.device != inp_meta.device
+                or ini_meta.device != inp_meta.device
+            ):
+                unimplemented(
+                    f"Expected the init, the input and the outputs of combine_fn to be a tensor on device {inp_meta.device} but "
+                    + f"got {ini_meta.device}, {inp_meta.device}, {carry_meta.device}, {out_meta.device}"
+                )
+            if carry_meta.dtype != ini_meta.dtype:
+                unimplemented(
+                    f"Expected the init and the new carry produced by the operator to be a tensor of {carry_meta.dtype} but "
+                    + f"got {ini_meta.dtype} and {carry_meta.dtype}"
+                )
+            if (
+                carry_meta.device != ini_meta.device
+                or carry_meta.dtype != ini_meta.dtype
+                or carry_meta.shape != ini_meta.shape
+            ):
+                unimplemented(
+                    f"Expected metadata of the combine_fn result {carry_meta} to be the same as "
+                    + f"the metadata of init with {ini_meta}"
+                )
+
+        combine_gm = torch.fx.GraphModule(dict(tx.output.nn_modules), combine_graph)
+        combine_fn_name = add_subgraph(tx, "scan_combine_fn", combine_gm)
+
+        p_args = (
+            make_attr(tx, combine_fn_name),
+            init_proxy,
+            input_proxy,
+            dim.as_proxy(),
+            reverse.as_proxy(),
+        )
+
+        with tx.fake_mode:
+            # For the fake mode, we need to duplicate the init tensor along the dim
+            # to have the same size as the input arguments
+            # We also do a clone with contiguous_format. This is to be consistent with
+            # eager semantic of map, which stacks the outputs. The result is contiguous
+            # as a result of the stack operation.
+            fake_out_shapes = [
+                tuple(
+                    -1 if i != dim_fake else dim_len
+                    for i, sh in enumerate(get_fake_value(o.as_proxy().node, tx).size())
+                )
+                for o in combine_result.items[1].items
+            ]
+            out_meta = (
+                [init_p.node.meta["example_value"].clone() for init_p in init_proxy],
+                list(
+                    t.as_proxy()
+                    .node.meta["example_value"]
+                    .expand(*sh)
+                    .clone(memory_format=torch.contiguous_format)
+                    for t, sh in zip(combine_result.items[1].items, fake_out_shapes)
+                ),
+            )
```

Comment:
Similar logic appears 3 times i think. Probably we should create a helper function for it.

Reply:
I consolidated all usecases and created a helper function [make_expanded_output_shape](https://github.com/pytorch/pytorch/blob/cb06c9cf487b748caa807bcc66207ce2f57f55cf/torch/_higher_order_ops/scan.py#L271-L273) for that in `scan.py`